### PR TITLE
Add a separate head for image classification.

### PR DESCRIPTION
**Pull request by jnphilipp**
Assumes that the dataset object, given to `train`, ..., has a function `load_image_class` similar to `load_mask` looking something like this:

```python
def load_image_class(self, image_id: int) -> int:
  # some logic to get image_class for a given image_id
  return image_class
```

### CODE REVIEW

File: mrcnn/model.py [hunk 1]
```diff
@@ -1269,6 +1269,10 @@ def load_image_gt(dataset, config, image_id, augment=False, augmentation=None,
     # Load image and mask
     image = dataset.load_image(image_id)
     mask, class_ids = dataset.load_mask(image_id)
+    if config.NUm_IMAGe_CLASSES:
```
Comment (bertsky) [inline mrcnn/model.py]:
```suggestion
    if config.NUM_IMAGE_CLASSES:
```

File: mrcnn/model.py [hunk 2]
```diff
@@ -1829,6 +1833,11 @@ def data_generator(self,image_ids):
                     batch_gt_masks = np.zeros(
                         (self.batch_size, gt_masks.shape[0], gt_masks.shape[1],
                         self.config.MAX_GT_INSTANCES), dtype=gt_masks.dtype)
+                    if self.config.NUm_IMAGe_CLASSES:
+                        batch_gt_image_classes = np.zeros(
+                            (self.batch_size, self.config.NUm_IMAGe_CLASSES),
```
Comment (bertsky) [inline mrcnn/model.py]:
```suggestion
                    if self.config.NUM_IMAGE_CLASSES:
                        batch_gt_image_classes = np.zeros(
                            (self.batch_size, self.config.NUM_IMAGE_CLASSES),
```

File: mrcnn/model.py [hunk 3]
```diff
@@ -1858,6 +1867,12 @@ def data_generator(self,image_ids):
                 batch_gt_class_ids[b, :gt_class_ids.shape[0]] = gt_class_ids
                 batch_gt_boxes[b, :gt_boxes.shape[0]] = gt_boxes
                 batch_gt_masks[b, :, :, :gt_masks.shape[-1]] = gt_masks
+                if self.config.NUm_IMAGe_CLASSES and gt_image_class:
+                    batch_gt_image_classes[b] = keras.utils.np_utils.to_categorical(
+                        gt_image_class,
+                        num_classes=self.config.NUm_IMAGe_CLASSES,
```
Comment (bertsky) [inline mrcnn/model.py]:
```suggestion
                if self.config.NUM_IMAGE_CLASSES and gt_image_class:
                    batch_gt_image_classes[b] = keras.utils.np_utils.to_categorical(
                        gt_image_class,
                        num_classes=self.config.NUM_IMAGE_CLASSES,
```

File: mrcnn/model.py [hunk 4]
```diff
@@ -1892,6 +1907,8 @@ def data_generator(self,image_ids):
                         batch_mrcnn_class_ids, -1)
                     outputs.extend(
                         [batch_mrcnn_class_ids, batch_mrcnn_bbox, batch_mrcnn_mask])
+                if self.config.NUm_IMAGe_CLASSES:
```
Comment (bertsky) [inline mrcnn/model.py]:
```suggestion
                if self.config.NUM_IMAGE_CLASSES:
```

File: mrcnn/model.py [hunk 5]
```diff
@@ -2134,6 +2151,13 @@ def build(self, mode, config):
             name="ROI",
             config=config)([rpn_class, rpn_bbox, anchors])
 
+        if config.NUm_IMAGe_CLASSES:
+            image_classes = KL.Dense(
+                config.NUm_IMAGe_CLASSES,
```
Comment (bertsky) [inline mrcnn/model.py]:
```suggestion
        if config.NUM_IMAGE_CLASSES:
            image_classes = KL.Dense(
                config.NUM_IMAGE_CLASSES,
```

File: mrcnn/model.py [hunk 6]
```diff
@@ -2198,14 +2222,16 @@ def build(self, mode, config):
                        mrcnn_class_logits, mrcnn_class, mrcnn_bbox, mrcnn_mask,
                        rpn_rois, output_rois,
                        rpn_class_loss, rpn_bbox_loss, class_loss, bbox_loss, mask_loss]
+            if config.NUm_IMAGe_CLASSES:
```
Comment (bertsky) [inline mrcnn/model.py]:
```suggestion
            if config.NUM_IMAGE_CLASSES:
```

File: mrcnn/model.py [hunk 7]
```diff
@@ -2223,12 +2249,21 @@ def build(self, mode, config):
                                               config.NUM_CLASSES,
                                               train_bn=config.TRAIN_BN)
 
-            model = KM.Model([input_image, input_image_meta, input_anchors],
-                             [detections, mrcnn_class, mrcnn_bbox,
-                              mrcnn_mask, rpn_rois, rpn_class, rpn_bbox,
-                              # output image_metas to facilitate unmolding
-                              KL.Lambda(lambda x: x)(input_image_meta)],
-                             name='mask_rcnn')
+            if config.NUm_IMAGe_CLASSES:
```
Comment (bertsky) [inline mrcnn/model.py]:
```suggestion
            if config.NUM_IMAGE_CLASSES:
```

Also, so we don't repeat ourselves, could we make that sth like:
```python
inputs = [...]
outputs = [...]
if config.NUM_IMAGE_CLASSES:
    outputs.append(image_class)
model = KM.Model(inputs, outputs, name='mask_rcnn')
```
?

File: mrcnn/model.py [hunk 8]
```diff
@@ -2134,6 +2151,13 @@ def build(self, mode, config):
             name="ROI",
             config=config)([rpn_class, rpn_bbox, anchors])
 
+        if config.NUm_IMAGe_CLASSES:
+            image_classes = KL.Dense(
```
Comment (bertsky) [inline mrcnn/model.py]: I recommend using the singular in the identifier here (`image_class`)

File: mrcnn/model.py [hunk 9]
```diff
@@ -2355,6 +2390,8 @@ def compile(self, learning_rate, momentum):
             for w in self.keras_model.trainable_weights
             if 'gamma' not in w.name and 'beta' not in w.name]
         self.keras_model.add_loss(tf.add_n(reg_losses))
+        if self.config.NUm_IMAGe_CLASSES:
```
Comment (bertsky) [inline mrcnn/model.py]:
```suggestion
        if self.config.NUM_IMAGE_CLASSES:
```

File: mrcnn/model.py [hunk 10]
```diff
@@ -2372,6 +2409,9 @@ def compile(self, learning_rate, momentum):
                 * self.config.LOSS_WEIGHTS.get(name, 1.))
             self.keras_model.add_metric(loss, name)
 
+        if self.config.NUm_IMAGe_CLASSES:
```
Comment (bertsky) [inline mrcnn/model.py]:
```suggestion
        if self.config.NUM_IMAGE_CLASSES:
```

File: mrcnn/model.py [hunk 11]
```diff
@@ -2198,14 +2222,16 @@ def build(self, mode, config):
                        mrcnn_class_logits, mrcnn_class, mrcnn_bbox, mrcnn_mask,
                        rpn_rois, output_rois,
                        rpn_class_loss, rpn_bbox_loss, class_loss, bbox_loss, mask_loss]
+            if config.NUm_IMAGe_CLASSES:
+                outputs.append(image_classes)
```
Comment (bertsky) [inline mrcnn/model.py]: I think we should place that output before the loss outputs. 

Also, I believe we need an additional loss output for `image_class` here explicitly. (See other comment about `add_loss` below.)

File: mrcnn/model.py [hunk 12]
```diff
@@ -2355,6 +2390,8 @@ def compile(self, learning_rate, momentum):
             for w in self.keras_model.trainable_weights
             if 'gamma' not in w.name and 'beta' not in w.name]
         self.keras_model.add_loss(tf.add_n(reg_losses))
+        if self.config.NUm_IMAGe_CLASSES:
+            self.keras_model.add_loss("categorical_crossentropy")
```
Comment (bertsky) [inline mrcnn/model.py]: How does this work? Network's `add_loss` can only take tensors AFAIK. 

Also, even if there's a std function like `keras.losses.SparseCategoricalCrossentropy()(gt_image_class, image_class)` (sparse since we are passing integers), it's better to define the loss graph as a custom function, because I'm pretty sure we'd also have to mask the batch-padded samples here – like we do in the other loss graphs, e.g. https://github.com/bertsky/Mask_RCNN/blob/876e42efb0b065006ef4daa59603fc96f1c4b19e/mrcnn/model.py#L1183-L1192

The place to define and add the loss is L2191ff (i.e. your L2215ff) rather than here.

**Review by bertsky — CHANGES_REQUESTED** [review]
Great work

File: mrcnn/model.py (hunk 12 above)
Comment (jnphilipp) [reply mrcnn/model.py]: I don't know why you think we pass the classes as integers [L1871-L1875](https://github.com/bertsky/Mask_RCNN/blob/461bbe34fa117df5979e197b19bb6ea7045ee0cc/mrcnn/model.py#L1871-L1875).

File: mrcnn/model.py (hunk 11 above)
Comment (jnphilipp) [reply mrcnn/model.py]: Done.

File: mrcnn/model.py (hunk 12 above)
Comment (bertsky) [reply mrcnn/model.py]: > I don't know why you think we pass the classes as integers [L1871-L1875](https://github.com/bertsky/Mask_RCNN/blob/461bbe34fa117df5979e197b19bb6ea7045ee0cc/mrcnn/model.py#L1871-L1875).

Right, non sparse. Also makes it easier to distinguish padded samples.

File: mrcnn/model.py [hunk 13]
```diff
@@ -1238,6 +1238,33 @@ def mrcnn_mask_loss_graph(target_masks, target_class_ids, pred_masks):
...
+    """Loss for the image classifier head of Mask RCNN.
+
+    target_image_class: [batch, num_image_classes]
+    pred_image_class: [batch, num_image_classes]. Uses zero padding to fill in the array.
+    """
+
+    non_zeros = tf.cast(tf.reduce_max(target_image_class, axis=1) > 0, 'float32')
+
+    # Loss
+    loss = tf.nn.softmax_cross_entropy_with_logits(
```
Comment (bertsky) [inline mrcnn/model.py]: We don't have logits here but probs. TF doc even says:
> Warning: This op expects unscaled logits, since it performs a softmax on logits internally for efficiency. Do not call this op with the output of softmax, as it will produce incorrect results.

I suggest `K.categorical_crossentropy(target_image_class, pred_image_class)`

**Review by bertsky — CHANGES_REQUESTED** [review]
Very good!

File: mrcnn/config.py [hunk 14]
```diff
@@ -72,6 +72,10 @@ class Config(object):
     # Number of classification classes (including background)
     NUM_CLASSES = 1  # Override in sub-classes
 
+    # Number of classes for a image classification. This is achieved through a separate
+    # head. If the values is None no separate head will be added.
+    NUM_IMAGE_CLASSES = None
```
Comment (bertsky) [inline mrcnn/config.py]:
```suggestion
    # Number of classes for a image classification. 
    # This is achieved by an additional head.
    # If the value is zero or None, then no separate head
    # (or additional outputs/losses) will be added.
    # (Needs a Dataset.load_image_class implementation.)
    NUM_IMAGE_CLASSES = 0
```

File: mrcnn/model.py [hunk 15]
```diff
@@ -2223,12 +2309,19 @@ def build(self, mode, config):
...
-                             name='mask_rcnn')
+            inputs = [input_image, input_image_meta, input_anchors]
+            outputs = [
+                detections,
+                mrcnn_class,
+                mrcnn_bbox,
+                mrcnn_mask,
+                rpn_rois,
+                rpn_class, rpn_bbox,# output image_metas to facilitate unmolding
+                KL.Lambda(lambda x: x)(input_image_meta)
```
Comment (bertsky) [inline mrcnn/model.py]:
```suggestion
                rpn_class, rpn_bbox,
                # output image_metas to facilitate unmolding
                KL.Lambda(lambda x: x)(input_image_meta)
```

**Review by bertsky — CHANGES_REQUESTED** [review]
Could you please add a default implementation (just printing a warning and returning `None`) for `utils.Dataset.load_image_class`, too?

**EDIT**: one remaining problem is that you still need to generalize the inference/detection handling for the additional-conditional output: It used to be just `unmold_outputs` on the full output tuple but with an explicit list of outputs in the fn def. But now there may or may not be an additional 9th output, depending on `self.config.NUM_IMAGE_CLASSES`. 

I therefore suggest rewriting `unmold_outputs` in the following way:
- take a single list-valued arg
- check its length, distinguish between 8 and 9:
- bind the first 8 members to the old arg names, and the 9th to a new `image_class` (i.e. `[batch, NUM_IMAGE_CLASSES]` array) or `None`
- in the `results`, unless `image_class == None`, add `"image_class": image_class[i]`

(We could of course also get the `np.argmax` right away, but perhaps the probability is of interest for the application, too.)

File: mrcnn/model.py [hunk 16]
```diff
@@ -3099,6 +3100,35 @@ def unmold_outputs(detections, mrcnn_class, mrcnn_bbox, mrcnn_mask,
...
+            mrcnn_mask,
+            rpn_rois,
+            rpn_class,
+            rpn_bbox,
+            image_metas,
+            image_class
+        ) = args
+    else:
+        print("Unrecognized number of arguments, needs to be either 8 or 9.")
+        return None
```
Comment (bertsky) [inline mrcnn/model.py]:
```suggestion
        raise ValueError("Invalid number of arguments, must be either 8 or 9.")
```

File: mrcnn/model.py [hunk 17]
```diff
@@ -3099,6 +3100,35 @@ def unmold_outputs(detections, mrcnn_class, mrcnn_bbox, mrcnn_mask,
...
+            detections,
+            mrcnn_class,
+            mrcnn_bbox,
+            mrcnn_mask,
+            rpn_rois,
+            rpn_class,
+            rpn_bbox,
+            image_metas,
+            image_class
+        ) = args
```
Comment (bertsky) [inline mrcnn/model.py]: You always need to set `image_class`, otherwise it will be undefined below.
```suggestion
    if len(args) in [8, 9]:
        (detections, mrcnn_class, mrcnn_bbox, mrcnn_mask,
         rpn_rois, rpn_class, rpn_bbox, image_metas, 
         image_class) = (args + [None])[:9]
```

File: mrcnn/model.py [hunk 18]
```diff
@@ -3114,8 +3144,12 @@ def unmold_outputs(detections, mrcnn_class, mrcnn_bbox, mrcnn_mask,
             "scores": final_scores,
             "masks": final_masks,
         })
+
+        if image_class:
+            results["image_class"] = image_class[i]
```
Comment (bertsky) [inline mrcnn/model.py]: This does not work. The function operates on batches, so `results` is a list of images. You must append to the dict in the loop:
```suggestion
            "image_class": image_class[i]
        })
```

**Review by bertsky — CHANGES_REQUESTED** [review]
One more time, please!

File: mrcnn/model.py (hunk 17 above)
Comment (jnphilipp) [reply mrcnn/model.py]: I'm setting it. But your code is shorter.

Comment (bertsky) [reply mrcnn/model.py]: Sry, wood from the trees...

File: mrcnn/model.py [hunk 19]
```diff
@@ -1933,9 +1935,8 @@ def data_generator(self,image_ids):
                         batch_mrcnn_class_ids, -1)
                     outputs.extend(
                         [batch_mrcnn_class_ids, batch_mrcnn_bbox, batch_mrcnn_mask])
-            if self.config.NUM_IMAGE_CLASSES:
-                inputs.append(batch_gt_image_classes)
-                outputs.append(batch_gt_image_classes)
+                    if self.config.NUM_IMAGE_CLASSES:
+                        outputs.append(batch_gt_image_classes)
```
Comment (bertsky) [inline mrcnn/model.py:1940]: I still don't understand why our new option should depend on whether or not we train without RPN.

File: mrcnn/model.py [hunk 20]
```diff
@@ -2192,7 +2192,7 @@ def build(self, mode, config):
                     config.NUM_IMAGE_CLASSES,
                     activation="softmax",
                     name="image_class"
-            )(P6)
+            )(KL.Flatten()(P6))
```
Comment (bertsky) [inline mrcnn/model.py:2196]: Yes, make sense!

File: mrcnn/model.py (hunk 19 above)
Comment (jnphilipp) [reply mrcnn/model.py:1940]: So, I had another look through the Code, as per [doc](https://github.com/bertsky/Mask_RCNN/blob/8f9193807c2e77d70a0ec623a367d80aa30c6eac/mrcnn/model.py#L1776-L1778) during training `outputs` should be an empty array.

The `detection_targets` is a debug variable to get some debug outputs, so our new output should only be there if the debug output is there.

And as per lines [L1842-L1844](https://github.com/bertsky/Mask_RCNN/blob/8f9193807c2e77d70a0ec623a367d80aa30c6eac/mrcnn/model.py#L1842-L1844) to generate the detection targets one needs the `rpn_rois` which only gets generated if `random_rois > 0`.

That said, we should consider whether it is at all necessary to add the `image_class` to the output at all.

Comment (bertsky) [reply mrcnn/model.py:1940]: > So, I had another look through the Code, as per [doc](https://github.com/bertsky/Mask_RCNN/blob/8f9193807c2e77d70a0ec623a367d80aa30c6eac/mrcnn/model.py#L1776-L1778) during training `outputs` should be an empty array.

Ah, yes, of course! I completely forgot about that. This implementation uses the `Model.add_loss` API exclusively, i.e. it passes `None` to the `loss` argument of `compile` for every output:
https://github.com/matterport/Mask_RCNN/blob/3deaec5d902d16e1daf56b62d5971d428dc920bc/mrcnn/model.py#L2188

So, indeed, in the normal training regime, we must not pass any outputs (because no loss must be computed from these outputs).

> The `detection_targets` is a debug variable to get some debug outputs, so our new output should only be there if the debug output is there.
> 
> And as per lines [L1842-L1844](https://github.com/bertsky/Mask_RCNN/blob/8f9193807c2e77d70a0ec623a367d80aa30c6eac/mrcnn/model.py#L1842-L1844) to generate the detection targets one needs the `rpn_rois` which only gets generated if `random_rois > 0`.
> 
> That said, we should consider whether it is at all necessary to add the `image_class` to the output at all.

I still think we should add the **tensor variable** to the output IMO. But in the generator, we need to keep the indentation (under the scope of `random_rois` and `detection_targets`), so it stays away from the output tuple.
